GenericList/string_list.h

```diff
@@ -1,31 +1,78 @@
 #pragma once
 
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+
 class List_String {
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
 public:
     List_String();
-    ~List_String();
 
-    int Capacity();
-    bool Capacity(int new_capacity);
-    int Count();
+    friend void swap(List_String& first, List_String& second);
+
+    //Rule of 3
+    ~List_String(); //Destructor
+    List_String(const List_String& other); //Copy Constructor
+    List_String& operator=(const List_String& other); //Copy assignement
+
+    //Rule of 5
+    //R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+    //basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+    List_String(List_String&& other);
+    List_String& operator=(List_String&& other);
+
+
+
+
+    int Capacity() const;
+    int Count() const;
+
+    //Sets the capacity of the internal array to new_capacity. If new_capacity is smaller than Count, do nothing.
+    void Capacity(int new_capacity);
+
+    std::string ToString() const;
 
-    std::string ToString();
+    void Add(const std::string& new_val);
+    void Add(std::string&& new_val);
 
-    bool Add(std::string new_val);
-    bool Contains(std::string val);
+    bool Contains(const std::string& val) const;
     bool RemoveAt(int index);
+
+    int IndexOf(const std::string& val) const;
+
+    //Returns true if a relevant element exists, and removes it.
+    bool Remove(const std::string& val);
+
+    const std::string& operator[](int index) const { return Get(index); }
+    std::string& operator[](int index) { return Get(index); }
```

Comment:
I'd just access the pointer directly here, if the user wants to do some exception  handling they can call `Get` directly instead.
```suggestion
    const std::string& operator[](int index) const { return data[index]; }
    std::string& operator[](int index) { return data[index]; }
```